File: src/plot_privacy_construction.py
#%%
import sys
import numpy as np
from plotnine import *
import pandas as pd
import matplotlib.pyplot as plt


if len(sys.argv) > 2:
    _args = sys.argv[1:]
else:
    _args = [
        "../output/analytics/base_analytics/departure-diffusion_exp/base_analytics_date_2019_04_08_d_2.csv",
        "../output/analytics/k_anonymous/departure-diffusion_exp/k_anonymous_analytics_date_2019_04_08_d_2.csv",
        "../output/analytics/sensitivity/GDP/GDP_analytics_s_10_e_1_k_NA_m_NA_date_2019_04_08_d_2.csv",
        "../output/analytics/sensitivity/CMS/CMS_analytics_s_10_e_5_k_20_m_2340_date_2019_04_08_d_2.csv",
        '../output/figs/k_anonymity_construction.png',
        '../output/figs/gdp_construction.png',
        '../output/figs/cms_construction.png',
    ]

_outputs = _args[-3:]

# %%
base = pd.read_csv(_args[0], dtype={'geoid_o': str, 'geoid_d': str})
k_anonymous = pd.read_csv(_args[1], dtype={'geoid_o': str, 'geoid_d': str}).rename(columns={'count': 'count_k_anonymous'})
gdp = pd.read_csv(_args[2], dtype={'geoid_o': str, 'geoid_d': str}).rename(columns={'count': 'count_gdp'})
cms = pd.read_csv(_args[3], dtype={'geoid_o': str, 'geoid_d': str}).rename(columns={'count': 'count_cms'})
# %%
base.sort_values(by='count', inplace=True, ascending=False)
base['id'] = range(1, base.shape[0] + 1)
# %%
base = pd.merge(base, k_anonymous, on=['geoid_o', 'geoid_d'], how='left',  validate='one_to_one')
base = pd.merge(base, gdp, on=['geoid_o', 'geoid_d'], how='left',  validate='one_to_one')
base = pd.merge(base, cms, on=['geoid_o', 'geoid_d'], how='left',  validate='one_to_one')

#%%
# Compute pull values for manuscript
# Suppression with k-anonymity

base_k_anonymous_suppressed = base[pd.isna(base['count_k_anonymous'])]

od_pairs_suppressed_k_anonymous = 1- (base_k_anonymous_suppressed.shape[0] / base.shape[0])
trips_suppressed_k_anonymous = 1- (base_k_anonymous_suppressed['count'].sum() / base['count'].sum())
print(f"K-anonymity retained: {od_pairs_suppressed_k_anonymous:.2%} OD pairs,  {trips_suppressed_k_anonymous:.2%} trips")

base['ape_count_gdp'] = np.abs(base['count'] - base['count_gdp']) / base['count']
base['ape_count_cms'] = np.abs(base['count'] - base['count_cms']) / base['count']

accuracy_threshold = 0.1

base_gdp_suppressed = base[base['ape_count_gdp'] > accuracy_threshold]

od_pairs_suppressed_gdp = 1-(base_gdp_suppressed.shape[0] / base.shape[0])
trips_suppressed_gdp = 1-(base_gdp_suppressed['count'].sum() / base['count'].sum())
print(f"GDP retained: {od_pairs_suppressed_gdp:.2%} OD pairs,  {trips_suppressed_gdp:.2%} trips")

base_cms_suppressed = base[base['ape_count_cms'] > accuracy_threshold]

od_pairs_suppressed_cms = 1-(base_cms_suppressed.shape[0] / base.shape[0])
trips_suppressed_cms = 1-(base_cms_suppressed['count'].sum() / base['count'].sum())
print(f"CMS retained: {od_pairs_suppressed_cms:.2%} OD pairs,  {trips_suppressed_cms:.2%} trips")

# truncate negative values to 0
gdp['count_gdp'] = gdp['count_gdp'].clip(lower=0)
cms['count_cms'] = cms['count_cms'].clip(lower=0)

#%%
k_anonymity_suppressed = base[pd.isna(base['count_k_anonymous'])]
print(f"K-anonymity suppressed: {1-(k_anonymity_suppressed.shape[0] / base.shape[0]):%} OD pairs,  {1-(k_anonymity_suppressed['count'].sum() / base['count'].sum()):%} trips")

# %%
base_long = pd.melt(base, id_vars=['id', 'geoid_o', 'geoid_d', 'count'], value_vars=['count_k_anonymous', 'count_gdp', 'count_cms'], var_name='mechanism')
# %%
start_exp=0
max_value = base_long['value'].max(skipna=True)
end_exp = np.log10(max_value) 

pa = (ggplot(base_long.loc[base_long['mechanism'] == 'count_k_anonymous']) + 
 geom_point(aes(x='id', y='count'), color='blue', size=0.1) + 
 geom_point(aes(x='id', y='value'), color='red', size=0.1) + 
 geom_hline(yintercept=10, color='black', linetype='dashed') +
 scale_y_continuous(trans='pseudo_log', limits=[base_long['value'].min(skipna=True), base_long['value'].max(skipna=True)],
                    breaks = [0] + [10**x for x in range(0, int(end_exp)+2)]) + 
 scale_x_continuous(trans='pseudo_log') + 
 labs(x = "Origin-destination pair (log scale)",
      y = "Number of trips (log scale)") + 
    theme_classic() + 
    theme(axis_text_x=element_blank(),
          axis_ticks_major_x=element_blank())
)
# %%
def plot_privacy(data, mechanism, ymin):

    return (ggplot(data.loc[base_long['mechanism'] == mechanism]) + 
        geom_point(aes(x='id', y='value'), color='red', size=0.1) + 
        geom_point(aes(x='id', y='count'), color='blue', size=0.1) + 
        scale_y_continuous(trans='pseudo_log', limits=[ymin, base_long['value'].max(skipna=True)],
                           breaks = [0] + [10**x for x in range(0, int(end_exp)+2)]) + 
        scale_x_continuous(trans='pseudo_log') + 
        labs(x = "Origin-destination pair (log scale)",
            y = "") + 
            theme_classic() + 
            theme(axis_text_x=element_blank(),
                axis_ticks_major_x=element_blank())
        )
# %%
pb = plot_privacy(base_long, 'count_gdp', ymin=0)
pc = plot_privacy(base_long, 'count_cms', ymin=0)
# %%
width = 4.5
height = 5

pa.save(_outputs[0], width=width, height=height, dpi=300)
pb.save(_outputs[1], width=width, height=height, dpi=300)
pc.save(_outputs[2], width=width, height=height, dpi=300)

# %%

File: src/plot_privacy_construction.R
suppressPackageStartupMessages({
  library(data.table)
  library(ggplot2)
})

if (interactive()) {
  .args <- c(
    "output/analytics/sensitivity/privacy_sensitivity_errors_date_2019_04_08_d_2.csv",
    "output/analytics/k_anonymous/departure-diffusion_exp/k_anonymous_analytics_date_2019_04_08_d_2.csv",
    "output/figs/construction_comparison.png"
  )
} else {
  .args <- commandArgs(trailingOnly = T)
}

.outputs <- tail(.args, 1)

errors <- fread(.args[1])

error_cols <- c('weight', 'squared_error',
                'weighted_squared_error', 'absolute_percentage_error',
                'weighted_absolute_percentage_error')
errors[, (error_cols) := NULL]
errors[, construction := ifelse(construction == "GDP", "CDP", construction)]
errors[, count := as.integer(count)]
errors[, count_private := as.integer(count_private)]

k_anon <- fread(.args[2])
k_anon[, count := as.integer(count)]
errors[, count_private := as.integer(count_private)]

# Select representative examples of the CDP and CMS mechanisms
errors_gdp <- subset(errors, 
                     construction == "CDP" & epsilon == 1 & sensitivity == 10
)

errors_cms <- subset(errors, 
                     construction == "CMS" & epsilon == 5 & sensitivity == 10 & 
                       k == 205 & m == 2340
)

errors_gdp_cms_comparison <- rbind(errors_gdp, errors_cms)
param_cols <- c('epsilon', 'sensitivity', 'm', 'k')
errors_gdp_cms_comparison[, (param_cols) := NULL]

# Compute origin-destination pair frequency rank based on true count
od_counts <- unique(errors[, .(geoid_o, geoid_d, count)])
od_counts <- od_counts[order(-count)]
od_counts[, id := .I]

# Format k-anonymity data for comparison
colnames(k_anon) <- c('geoid_o', 'geoid_d', 'count_private')

k_anon_full <- od_counts[, .(geoid_o, geoid_d, count)]
k_anon <- k_anon_full[k_anon, on=c('geoid_o', 'geoid_d'), count_private := count_private]
k_anon[, construction := 'K-anonymity']

# Combine all three privacy mechanisms
errors_comparison <- rbind(errors_gdp_cms_comparison, k_anon)

# Assign frequency ranks
errors_comparison[od_counts, on=c("geoid_o", "geoid_d"), id := id]
k_anon[od_counts, on=c("geoid_o", "geoid_d"), id := id]

# Pivot data to compare private counts with true counts
errors_comparison_long <- melt(errors_comparison, 
     id.vars = c("geoid_o", "geoid_d", "construction", "id"),
     variable.name = "privacy")

errors_comparison_long[, construction := factor(
  construction,
  levels=c("K-anonymity", "CDP", "CMS"),
  labels=c("K-anonymity",
           "Central DP",
           "Local DP (CMS)"))]

k_anon[, construction := factor(
  construction,
  levels=c("K-anonymity", "CDP", "CMS"),
  labels=c("K-anonymity",
           "Central DP",
           "Local DP (CMS)"))]

errors_comparison_long[, privacy := factor(
  privacy,
  levels=c("count", "count_private"),
  labels=c("True value",
           "Privatised value"))]

k_anon[, privacy := factor(
  "count_private",
  levels=c("count_private"),
  labels=c("Privatised value"))]

p <- ggplot(subset(errors_comparison_long, !is.na(value))) + 
  geom_point(aes(x = id, y = pmax(value, 0), color=privacy),
             size=0.1) + 
  geom_point(data=k_anon, aes(x = id, y=count_private, color=privacy),
             size=0.1) + 
  geom_hline(data = subset(errors_comparison_long, construction == "K-anonymity" & !is.na(value)),
             aes(yintercept = 10), color = "black", linetype='dashed', size=0.2) +
  facet_wrap(~construction, nrow=2) + 
  scale_y_continuous(trans="log10", labels = scales::comma,
                     breaks=c(0.1, 1, 10, 100, 1000, 10000)) + 
  scale_color_manual(values = c("Privatised value" = "red", "True value" = "blue")) +
  theme_classic() + 
  theme(legend.position = c(0.75, 0.22),
        strip.background = element_rect(size=0),
        strip.text = element_text(face='bold')) + 
  scale_x_continuous(trans="log10", 
                     breaks=c(1, max(errors_comparison_long$id)),
                     labels = c("High", "Low")) + 
  labs(color="Privacy",
       x = expression(paste("Origin-destination pair frequency rank (", log[10], " scale)")),
       y = expression(paste("Number of trips (", log[10], " scale)")))

ggsave(.outputs[1],
       p,
       width=8,
       height=6, 
       units="in")